account-kit/smart-contracts/src/ma-v2/account/common/modularAccountV2Base.ts

```diff
@@ -0,0 +1,217 @@
+import {
+  createBundlerClient,
+  getEntryPoint,
+  InvalidEntityIdError,
+  InvalidNonceKeyError,
+  type AccountOp,
+  type EntryPointDef,
+  type SmartAccountSigner,
+  type SmartContractAccountWithSigner,
+  type ToSmartContractAccountParams,
+} from "@aa-sdk/core";
+import { DEFAULT_OWNER_ENTITY_ID } from "../../utils.js";
+import {
+  type Hex,
+  type Address,
+  type Chain,
+  type Transport,
+  encodeFunctionData,
+  maxUint32,
+  zeroAddress,
+  getContract,
+  concatHex,
+  maxUint152,
+} from "viem";
+import { modularAccountAbi } from "../../abis/modularAccountAbi.js";
+import { serializeModuleEntity } from "../../actions/common/utils.js";
+
+export const executeUserOpSelector: Hex = "0x8DD7712F";
+
+export type SignerEntity = {
+  isGlobalValidation: boolean;
+  entityId: number;
+};
+
+export type ExecutionDataView = {
+  module: Address;
+  skipRuntimeValidation: boolean;
+  allowGlobalValidation: boolean;
+  executionHooks: readonly Hex[];
+};
+
+export type ValidationDataView = {
+  validationHooks: readonly Hex[];
+  executionHooks: readonly Hex[];
+  selectors: readonly Hex[];
+  validationFlags: number;
+};
+
+export type ValidationDataParams =
+  | {
+      validationModuleAddress: Address;
+      entityId?: never;
+    }
+  | {
+      validationModuleAddress?: never;
+      entityId: number;
+    };
+
+export type MAV2Account<
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+> = SmartContractAccountWithSigner<"MAV2Account", TSigner, "0.7.0"> & {
```

Comment:
I think this base type should probably be generic on `source` as well since `source` should be unique per instance (ie. the first generic parameter in `SmartContractAccountWithSigner`)